Migrate ERC20 to component (#780)
OpenZeppelin/cairo-contracts: docs/modules/ROOT/pages/erc20.adoc

:dual-interfaces: xref:/interfaces.adoc#dual_interfaces[Dual interfaces]
:erc20-supply: xref:/guides/erc20-supply.adoc[Creating ERC20 Supply]

The following interface represents the full ABI of the Contracts for Cairo `ERC20` component.
The interface includes the `IERC20` standard interface as well as non-standard functions like `increase_allowance`.

> To support older ERC20 token deployments, as mentioned in {dual-interfaces}, the `ERC20` component also includes the previously defined functions written in camelCase.

Reviewer: ```suggestion
To support older ERC20 token deployments, as mentioned in {dual-interfaces}, the `ERC20` component also includes an implementation of the interface written in camelCase.
```
the current version suggests that the functions are there by default (as it was with the preset)

----
#[starknet::contract]
mod MyToken {
    use openzeppelin::token::erc20::ERC20 as erc20_component;
    use openzeppelin::token::erc20::interface;
    use starknet::ContractAddress;

    component!(path: erc20_component, storage: erc20, event: ERC20Event);

    #[abi(embed_v0)]
    impl ERC20Impl = erc20_component::ERC20Impl<ContractState>;
    #[abi(embed_v0)]
    impl ERC20MetadataImpl = erc20_component::ERC20MetadataImpl<ContractState>;
    impl ERC20InternalImpl = erc20_component::InternalImpl<ContractState>;

    #[storage]
    struct Storage {
        #[substorage(v0)]
        erc20: erc20_component::Storage
    }

    #[event]
    #[derive(Drop, starknet::Event)]
    enum Event {
        ERC20Event: erc20_comp::Event
    }

    #[constructor]
    fn constructor(
        ref self: ContractState,
        initial_supply: u256,
        recipient: ContractAddress
    ) {
        let name = 'MyToken';
        let symbol = 'MTK';

        self.erc20.initializer(name, symbol);
        self.erc20._mint(recipient, initial_supply);
    }
}
----

In order for the `MyToken` contract to integrate the `ERC20` component, it instantiates both the `ERC20Impl` and `ERC20MetadataImpl`.

> The `ERC20Impl` instance embeds the implementation and exposes the `IERC20` interface.

Reviewer: isn't actually the `#[abi(embed_v0)]` directive the one that embeds the impl?

Author: Yeah, I took the directive for granted here. I'll update to make this more accurate

Here's what that looks like:

[,javascript]
----
#[starknet::contract]
mod MyToken {
    use openzeppelin::token::erc20::ERC20 as erc20_component;
    use openzeppelin::token::erc20::interface;
    use starknet::ContractAddress;

    component!(path: erc20_component, storage: erc20, event: ERC20Event);

    #[abi(embed_v0)]
    impl ERC20Impl = erc20_component::ERC20Impl<ContractState>;
    #[abi(embed_v0)]
    impl ERC20MetadataImpl = erc20_component::ERC20MetadataImpl<ContractState>;
    impl ERC20InternalImpl = erc20_component::InternalImpl<ContractState>;

    #[storage]
    struct Storage {
        #[substorage(v0)]
        erc20: erc20_component::Storage
    }

    #[event]
    #[derive(Drop, starknet::Event)]
    enum Event {
        ERC20Event: erc20_comp::Event
    }

    #[constructor]
    fn constructor(
        ref self: ContractState,
        initial_supply: u256,
        recipient: ContractAddress
    ) {
        let name = 'MyToken';
        let symbol = 'MTK';

        self.erc20.initializer(name, symbol);
        self.erc20._mint(recipient, initial_supply);
    }
}
----

> In order for the `MyToken` contract to integrate the ERC20 component, it instantiates both the `ERC20Impl` and `ERC20MetadataImpl`.

Author: Agreed, it is inconsistent. I edited the doc so that the general ERC20 references don't include back ticks but the specific `ERC20` refs do. Let me know what you think

> In order for the `MyToken` contract to integrate the ERC20 component, it instantiates both the `ERC20Impl` and `ERC20MetadataImpl`.

Author: For general, I mean not referencing anything specific to Contracts for Cairo e.g. "To support older ERC20 token deployments..."

And for specific, I mean referencing the Contracts for Cairo ERC20 mod/impl/interface e.g. "The `ERC20` component includes..."

> In order for the `MyToken` contract to integrate the ERC20 component, it instantiates both the `ERC20Impl` and `ERC20MetadataImpl`.

Reviewer: to me the backticks only represents code entities, and without i'd interpret it as the ERC20 standard. I think this is close to what you mean, but I don't think readers will catch the OZ/non-OZ distinction.

> In order for the `MyToken` contract to integrate the ERC20 component, it instantiates both the `ERC20Impl` and `ERC20MetadataImpl`.

Author: > but I don't think readers will catch the OZ/non-OZ distinction

Fair enough, I'll edit and improve the content

}
----

> `MyToken` instantiates both the `ERC20Impl` and `ERC20MetadataImpl` with the embed directive which integrates the implementations into the contract.

Reviewer: ```suggestion
`MyToken` integrates both the `ERC20Impl` and `ERC20MetadataImpl` with the embed directive which marks the implementations as external in the contract.
```
i'm unsure about using this word, it's not actually an instance isn't it? also wanted to make sure we say they're exposed (integrating the impl could be without exposing it)

Author: > i'm unsure about using this word, it's not actually an instance isn't it?

I suppose it depends on how we define instance. IMO a component impl with a generic state that gets actualized with a specific contract state becomes an instantiation. No issue with using "integrates" though

> also wanted to make sure we say they're exposed (integrating the impl could be without exposing it)

Yes, good call

:eip-discussion: https://github.com/ethereum/EIPs/issues/724[EIP discussion]

Cairo, like Solidity, does not support {floating-point}.
To get around this limitation, ERC20 token contracts may offer a `decimals` field which communicates to outside interfaces (wallets, exchanges, etc.) how the token should be displayed.
For instance, suppose a token had a `decimals` value of `3` and the total token supply was `1234`.
An outside interface would display the token supply as `1.234`.
In the actual contract, however, the supply would still be the integer `1234`.
In other words, *the decimals field in no way changes the actual arithmetic* because all operations are still performed on integers.

Most contracts use `18` decimals and this was even proposed to be compulsory (see the {eip-discussion}).

> The Contracts for Cairo `ERC20` component includes the `decimals` function that returns `18` by default to save on gas fees.

Reviewer: ```suggestion
The Contracts for Cairo `ERC20` component includes a `decimals` function that returns `18` by default to save on gas fees.
```

----
#[starknet::contract]
mod MyToken {
    use openzeppelin::token::erc20::ERC20 as erc20_component;
    use openzeppelin::token::erc20::interface;
    use starknet::ContractAddress;

    component!(path: erc20_component, storage: erc20, event: ERC20Event);

    #[abi(embed_v0)]
    impl ERC20Impl = erc20_component::ERC20Impl<ContractState>;
    impl ERC20InternalImpl = erc20_component::InternalImpl<ContractState>;

    #[storage]
    struct Storage {
        #[substorage(v0)]
        erc20: erc20_component::Storage,
        // The decimals value is stored locally

>         MyToken_decimals: u8

Reviewer: why prefixing? nothing bad i guess, but that's for components -- as long as you're using components following the pattern

Author: Right, force of habit

        let name = 'MyToken';
        let symbol = 'MTK';

        self.erc20.initializer(name, symbol);
        self.erc20._mint(recipient, initial_supply);
    }

    #[external(v0)]
    impl ERC20MetadataImpl of interface::IERC20Metadata<ContractState> {
        fn name(self: @ContractState) -> felt252 {
            self.erc20.name()
        }

        fn symbol(self: @ContractState) -> felt252 {
            self.erc20.symbol()
        }

        fn decimals(self: @ContractState) -> u8 {
            self.MyToken_decimals.read()
        }
    }

    #[generate_trait]
    impl InternalImpl of InternalTrait {
        fn _set_decimals(ref self: ContractState, decimals: u8) {
            self.MyToken_decimals.write(decimals);
        }
    }
}
----

This contract expects a `decimals` argument in the constructor and uses an internal function to write the decimals to storage.

> Note that the `MyToken_decimals` state variable must be stored in the local contract's storage because this variable does not exist in the Contracts for Cairo library.

Reviewer: ```suggestion
Note that the `MyToken_decimals` state variable must be defined in the contract's storage because this variable does not exist in the component offered by OpenZeppelin Contracts for Cairo.
```
